File: scripts/migrate-public-r2-catalog.mjs
import fs from "node:fs";
import https from "node:https";
import path from "node:path";
import dotenv from "dotenv";
import {
  CopyObjectCommand,
  GetObjectCommand,
  HeadObjectCommand,
  ListObjectsV2Command,
  PutObjectCommand,
  S3Client,
} from "@aws-sdk/client-s3";
import { NodeHttpHandler } from "@smithy/node-http-handler";

// Load the checked-in local settings last, so the current local R2 token wins.
const env = {
  ...(fs.existsSync(".env") ? dotenv.parse(fs.readFileSync(".env")) : {}),
  ...(fs.existsSync(".env.local") ? dotenv.parse(fs.readFileSync(".env.local")) : {}),
};

const required = ["R2_ACCOUNT_ID", "R2_ACCESS_KEY_ID", "R2_SECRET_ACCESS_KEY"];
const missing = required.filter((name) => !env[name]);
if (missing.length) throw new Error(`R2_ENV_MISSING:${missing.join(",")}`);

const sourceBucket = env.R2_PRIVATE_BUCKET_NAME || env.R2_BUCKET_NAME || env.R2_BUCKET;
const publicBucket = env.R2_PUBLIC_BUCKET_NAME;
const publicBaseUrl = String(env.R2_PUBLIC_BASE_URL || env.R2_PUBLIC_URL || "").replace(/\/$/, "");
if (!sourceBucket || !publicBucket || !publicBaseUrl) {
  throw new Error("R2_BUCKET_OR_PUBLIC_URL_MISSING");
}

const client = new S3Client({
  region: "auto",
  endpoint: `https://${env.R2_ACCOUNT_ID}.r2.cloudflarestorage.com`,
  requestHandler: new NodeHttpHandler({
    // Avoid the Windows/OpenSSL TLS-session error seen during bulk CopyObject calls.
    httpsAgent: new https.Agent({ keepAlive: false, maxCachedSessions: 0, maxSockets: 1 }),
  }),
  credentials: {
    accessKeyId: env.R2_ACCESS_KEY_ID,
    secretAccessKey: env.R2_SECRET_ACCESS_KEY,
  },
});

async function request(command) {
  let lastError;
  for (let attempt = 1; attempt <= 5; attempt += 1) {
    try {
      return await client.send(command);
    } catch (error) {
      lastError = error;
      if (attempt === 5) break;
      await new Promise((resolve) => setTimeout(resolve, attempt * 750));
    }
  }
  throw lastError;
}

const masterPath = path.resolve("private-data/images-master.json");
const fallbackManifestPath = path.resolve("public/data/images-public.json");
const master = JSON.parse(fs.readFileSync(masterPath, "utf8"));
const images = Array.isArray(master.images) ? master.images : [];
if (images.length !== 1583 || new Set(images.map((image) => image.id)).size !== 1583) {
  throw new Error("MASTER_COUNT_OR_IDS_INVALID");
}

const freeImages = images.filter((image) => image.plan_type === "free");
if (freeImages.length !== 37) throw new Error("FREE_IMAGE_COUNT_INVALID");
const args = new Set(process.argv.slice(2));

function extensionFor(key) {
  const match = String(key || "").match(/\.([a-z0-9]+)$/i);
  if (!match || !/^(jpg|jpeg|png|webp)$/i.test(match[1])) throw new Error(`UNSAFE_IMAGE_EXTENSION:${key}`);
  return match[1].toLowerCase();
}

function publicUrlFor(key) {
  return `${publicBaseUrl}/${key.split("/").map(encodeURIComponent).join("/")}`;
}

function copySourceFor(key) {
  return `/${sourceBucket}/${key.split("/").map(encodeURIComponent).join("/")}`;
}

async function copy(sourceKey, destinationKey, contentType) {
  try {
    await request(new HeadObjectCommand({ Bucket: publicBucket, Key: destinationKey }));
    return "existing";
  } catch (error) {
    if (error?.name !== "NotFound" && error?.$metadata?.httpStatusCode !== 404) throw error;
  }
  await request(new CopyObjectCommand({
    Bucket: publicBucket,
    Key: destinationKey,
    CopySource: copySourceFor(sourceKey),
    ContentType: contentType,
    MetadataDirective: "REPLACE",
    CacheControl: "public, max-age=31536000, immutable",
  }));
  return "copied";
}

async function runPool(items, worker, concurrency = 3) {
  let index = 0;
  const workers = Array.from({ length: Math.min(concurrency, items.length) }, async () => {
    while (true) {
      const current = index++;
      if (current >= items.length) return;
      await worker(items[current]);
    }
  });
  await Promise.all(workers);
}

const freePublicKeys = new Map(
  freeImages.map((image) => [image.id, `free/originals/${image.id}.${extensionFor(image.original_key)}`]),
);
const migrationAssets = [
  ...images.map((image) => ({
    sourceKey: image.thumbnail_key,
    destinationKey: image.thumbnail_key,
    contentType: "image/webp",
  })),
  ...freeImages.map((image) => ({
    sourceKey: image.original_key,
    destinationKey: freePublicKeys.get(image.id),
    contentType: `image/${extensionFor(image.original_key) === "jpg" ? "jpeg" : extensionFor(image.original_key)}`,
  })),
];

const batchArg = process.argv.find((arg) => arg.startsWith("--batch="));
if (batchArg) {
  const [offsetText, limitText = "10"] = batchArg.slice("--batch=".length).split(":");
  const offset = Number(offsetText);
  const limit = Number(limitText);
  if (!Number.isInteger(offset) || !Number.isInteger(limit) || offset < 0 || limit < 1) {
    throw new Error("BATCH_ARGUMENT_INVALID");
  }
  const batch = migrationAssets.slice(offset, offset + limit);
  await runPool(batch, async (asset) => {
    await copy(asset.sourceKey, asset.destinationKey, asset.contentType);
  }, 1);
  console.log(JSON.stringify({ batchOffset: offset, copied: batch.length, totalAssets: migrationAssets.length }));
  process.exit(0);
}

const thumbnailResults = { copied: 0, existing: 0 };
if (!args.has("--finalize")) {
  await runPool(images, async (image) => {
    if (!image.thumbnail_key || !String(image.thumbnail_key).startsWith("thumbnails/")) {
      throw new Error(`THUMBNAIL_KEY_INVALID:${image.id}`);
    }
    const status = await copy(image.thumbnail_key, image.thumbnail_key, "image/webp");
    thumbnailResults[status] += 1;
  });
}

const freeResults = { copied: 0, existing: 0 };
if (!args.has("--finalize")) {
  await runPool(freeImages, async (image) => {
    const extension = extensionFor(image.original_key);
    const key = freePublicKeys.get(image.id);
    const status = await copy(image.original_key, key, `image/${extension === "jpg" ? "jpeg" : extension}`);
    freeResults[status] += 1;
  });
}

const publicImages = images.map((image) => {
  const planType = image.plan_type === "free" ? "free" : "bundle";
  const item = {
    id: image.id,
    title: image.title || "圖片素材",
    category: image.category || image.category_id || "other",
    category_id: image.category_id || image.category || "other",
    category_name: image.category_name || "其他素材",
    thumbnail_url: publicUrlFor(image.thumbnail_key),
    preview_url: publicUrlFor(image.thumbnail_key),
    plan_type: planType,
  };
  if (planType === "free") item.download_url = publicUrlFor(freePublicKeys.get(image.id));
  return item;
});

const publicJson = JSON.stringify(publicImages, null, 2) + "\n";
const forbidden = /original_key|original_url|R2_SECRET|R2_ACCESS|private\//i;
if (forbidden.test(publicJson)) throw new Error("PUBLIC_MANIFEST_LEAK_BLOCKED");

await request(new PutObjectCommand({
  Bucket: publicBucket,
  Key: "catalog/images-public.json",
  Body: publicJson,
  ContentType: "application/json; charset=utf-8",
  CacheControl: "no-cache",
}));

// Retain the exact same sanitized catalog as the emergency Vercel fallback.
fs.writeFileSync(fallbackManifestPath, publicJson, "utf8");

async function listImageCount(prefix) {
  let continuationToken;
  let count = 0;
  do {
    const page = await request(new ListObjectsV2Command({
      Bucket: publicBucket,
      Prefix: prefix,
      ContinuationToken: continuationToken,
    }));
    count += (page.Contents || []).filter((item) => /\.(jpg|jpeg|png|webp)$/i.test(item.Key || "")).length;
    continuationToken = page.IsTruncated ? page.NextContinuationToken : undefined;
  } while (continuationToken);
  return count;
}

const publicThumbnailCount = await listImageCount("thumbnails/");
const publicFreeCount = await listImageCount("free/originals/");

const result = {
  publicManifestCount: publicImages.length,
  publicThumbnailCount,
  publicFreeOriginalCount: publicFreeCount,
  publicBundleOriginalCount: 0,
  thumbnailCopies: thumbnailResults,
  freeOriginalCopies: freeResults,
};
console.log(JSON.stringify(result));

File: scripts/migrate_public_r2_catalog.py
"""Copy only the public RxV catalog assets between two R2 buckets.

This intentionally never lists or changes source objects.  It uses the existing
master as the allow-list: thumbnails for all images and originals for free images.
"""
from concurrent.futures import ThreadPoolExecutor, as_completed
from datetime import datetime, timezone
from hashlib import sha256
import hmac
import json
from pathlib import Path
from urllib.error import HTTPError, URLError
from urllib.parse import quote
from urllib.request import Request, urlopen
import os
import time


ROOT = Path(__file__).resolve().parents[1]


def parse_env(path):
    values = {}
    if not path.exists():
        return values
    for line in path.read_text(encoding="utf-8").splitlines():
        line = line.strip()
        if not line or line.startswith("#") or "=" not in line:
            continue
        key, value = line.split("=", 1)
        values[key.strip()] = value.strip().strip('"').strip("'")
    return values


ENV = {**parse_env(ROOT / ".env"), **parse_env(ROOT / ".env.local")}
for name in ("R2_ACCOUNT_ID", "R2_ACCESS_KEY_ID", "R2_SECRET_ACCESS_KEY"):
    if not ENV.get(name):
        raise SystemExit(f"R2_ENV_MISSING:{name}")

SOURCE_BUCKET = ENV.get("R2_PRIVATE_BUCKET_NAME") or ENV.get("R2_BUCKET_NAME") or ENV.get("R2_BUCKET")
PUBLIC_BUCKET = ENV.get("R2_PUBLIC_BUCKET_NAME")
# This must be the new public bucket's Development URL (never the S3 API endpoint
# or the legacy bucket's public URL).
PUBLIC_BASE = (ENV.get("R2_PUBLIC_ASSET_URL") or "").rstrip("/")
if not SOURCE_BUCKET or not PUBLIC_BUCKET:
    raise SystemExit("R2_BUCKET_MISSING")

HOST = f"{ENV['R2_ACCOUNT_ID']}.r2.cloudflarestorage.com"
ENDPOINT = f"https://{HOST}"
EMPTY_HASH = sha256(b"").hexdigest()


def hmac_sha256(key, message):
    return hmac.new(key, message.encode("utf-8"), sha256).digest()


def signed_request(method, key, *, body=b"", extra_headers=None):
    now = datetime.now(timezone.utc)
    stamp = now.strftime("%Y%m%dT%H%M%SZ")
    date = now.strftime("%Y%m%d")
    payload_hash = sha256(body).hexdigest()
    headers = {
        "host": HOST,
        "x-amz-content-sha256": payload_hash,
        "x-amz-date": stamp,
        **(extra_headers or {}),
    }
    canonical_uri = "/" + quote(key, safe="/-_.~")
    canonical_headers = "".join(f"{name}:{' '.join(str(headers[name]).strip().split())}\n" for name in sorted(headers))
    signed_headers = ";".join(sorted(headers))
    scope = f"{date}/auto/s3/aws4_request"
    canonical = "\n".join([method, canonical_uri, "", canonical_headers, signed_headers, payload_hash])
    string_to_sign = "\n".join(["AWS4-HMAC-SHA256", stamp, scope, sha256(canonical.encode()).hexdigest()])
    signing_key = hmac_sha256(
        hmac_sha256(hmac_sha256(hmac_sha256(("AWS4" + ENV["R2_SECRET_ACCESS_KEY"]).encode(), date), "auto"), "s3"),
        "aws4_request",
    )
    signature = hmac.new(signing_key, string_to_sign.encode(), sha256).hexdigest()
    headers["Authorization"] = (
        f"AWS4-HMAC-SHA256 Credential={ENV['R2_ACCESS_KEY_ID']}/{scope}, "
        f"SignedHeaders={signed_headers}, Signature={signature}"
    )
    return Request(ENDPOINT + canonical_uri, data=body if method in ("PUT", "POST") else None,
                   headers=headers, method=method)


def send(method, key, *, body=b"", extra_headers=None, attempts=6):
    last_error = None
    for attempt in range(attempts):
        try:
            with urlopen(signed_request(method, key, body=body, extra_headers=extra_headers), timeout=45) as response:
                return response.status, dict(response.headers.items())
        except (HTTPError, URLError, OSError) as error:
            last_error = error
            if attempt == attempts - 1:
                break
            time.sleep(0.6 * (attempt + 1))
    raise last_error


master = json.loads((ROOT / "private-data" / "images-master.json").read_text(encoding="utf-8"))
images = master.get("images", [])
if len(images) != 1583 or len({item.get("id") for item in images}) != 1583:
    raise SystemExit("MASTER_COUNT_OR_IDS_INVALID")
free_images = [item for item in images if item.get("plan_type") == "free"]
if len(free_images) != 37:
    raise SystemExit("FREE_IMAGE_COUNT_INVALID")


def fallback_catalog():
    output = []
    for item in images:
        plan_type = "free" if item.get("plan_type") == "free" else "bundle"
        result = {
            "id": item["id"],
            "title": item.get("title") or "圖片素材",
            "category": item.get("category") or item.get("category_id") or "other",
            "category_id": item.get("category_id") or item.get("category") or "other",
            "category_name": item.get("category_name") or "其他素材",
            "thumbnail_url": f"/api/main?action=get-r2-image-thumbnail&id={item['id']}",
            "preview_url": f"/api/main?action=get-r2-image-thumbnail&id={item['id']}",
            "plan_type": plan_type,
        }
        if plan_type == "free":
            result["download_url"] = f"/api/main?action=get-r2-free-image-download&id={item['id']}"
        output.append(result)
    return output


if "--fallback-only" in os.sys.argv:
    (ROOT / "public" / "data" / "images-public.json").write_text(
        json.dumps(fallback_catalog(), ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
    )
    print(json.dumps({"fallback_manifest_count": len(images)}))
    raise SystemExit(0)

if not PUBLIC_BASE or ".r2.cloudflarestorage.com" in PUBLIC_BASE:
    raise SystemExit("R2_PUBLIC_ASSET_URL_MISSING_OR_NOT_PUBLIC")


def suffix(key):
    extension = key.rsplit(".", 1)[-1].lower() if "." in key else ""
    if extension not in {"jpg", "jpeg", "png", "webp"}:
        raise ValueError(f"UNSAFE_IMAGE_EXTENSION:{key}")
    return extension


free_keys = {item["id"]: f"free/originals/{item['id']}.{suffix(item['original_key'])}" for item in free_images}
assets = []
for item in images:
    source = item.get("thumbnail_key")
    if not source or not (source.startswith("thumbnails/") or source.startswith("images/thumbnails/")):
        raise SystemExit(f"THUMBNAIL_KEY_INVALID:{item.get('id')}")
    assets.append((source, source, "image/webp"))
for item in free_images:
    ext = suffix(item["original_key"])
    assets.append((item["original_key"], free_keys[item["id"]], f"image/{'jpeg' if ext == 'jpg' else ext}"))


def copy_asset(asset):
    source, destination, content_type = asset
    copy_source = "/" + SOURCE_BUCKET + "/" + quote(source, safe="/-_.~")
    send("PUT", f"{PUBLIC_BUCKET}/{destination}", extra_headers={
        "content-type": content_type,
        "cache-control": "public, max-age=31536000, immutable",
        "x-amz-copy-source": copy_source,
        "x-amz-metadata-directive": "REPLACE",
    })
    return destination


if "--manifest-only" not in os.sys.argv:
    completed = 0
    with ThreadPoolExecutor(max_workers=8) as pool:
        futures = [pool.submit(copy_asset, asset) for asset in assets]
        for future in as_completed(futures):
            future.result()
            completed += 1
            if completed % 100 == 0:
                print(f"COPIED={completed}", flush=True)


def public_url(key):
    return PUBLIC_BASE + "/" + "/".join(quote(part, safe="-_.~") for part in key.split("/"))


public_images = []
for item in images:
    plan_type = "free" if item.get("plan_type") == "free" else "bundle"
    public_item = {
        "id": item["id"],
        "title": item.get("title") or "圖片素材",
        "category": item.get("category") or item.get("category_id") or "other",
        "category_id": item.get("category_id") or item.get("category") or "other",
        "category_name": item.get("category_name") or "其他素材",
        "thumbnail_url": public_url(item["thumbnail_key"]),
        "preview_url": public_url(item["thumbnail_key"]),
        "plan_type": plan_type,
    }
    if plan_type == "free":
        public_item["download_url"] = public_url(free_keys[item["id"]])
    public_images.append(public_item)

manifest = (json.dumps(public_images, ensure_ascii=False, indent=2) + "\n").encode("utf-8")
if any(token in manifest.decode("utf-8").lower() for token in ("original_key", "original_url", "r2_secret", "r2_access", "private/")):
    raise SystemExit("PUBLIC_MANIFEST_LEAK_BLOCKED")

send("PUT", f"{PUBLIC_BUCKET}/catalog/images-public.json", body=manifest, extra_headers={
    "content-type": "application/json; charset=utf-8",
    "cache-control": "no-cache",
})
(ROOT / "public" / "data" / "images-public.json").write_text(
    json.dumps(fallback_catalog(), ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
)

print(json.dumps({
    "public_manifest_count": len(public_images),
    "public_thumbnail_count": len(images),
    "public_free_original_count": len(free_images),
    "public_bundle_original_count": 0,
}))
